src/model/CoilHeatingElectricMultiStage.cpp

```diff
@@ -0,0 +1,221 @@
+
+      auto t_stages = stages();
+      for (auto stage : t_stages) {
+        auto stageClone = stage.clone(model).cast<CoilHeatingElectricMultiStageStageData>();
+        t_clone.addStage(stageClone);
+      }
+      return t_clone;
+    }
+
+    std::vector<ModelObject> CoilHeatingElectricMultiStage_Impl::children() const {
+      return subsetCastVector<ModelObject>(stages());
+    }
+
+    std::vector<CoilHeatingElectricMultiStageStageData> CoilHeatingElectricMultiStage_Impl::stages() const {
+      std::vector<CoilHeatingElectricMultiStageStageData> result;
+      auto groups = extensibleGroups();
+      for (auto group : groups) {
+        auto target = group.cast<WorkspaceExtensibleGroup>().getTarget(OS_Coil_Heating_Electric_MultiStageExtensibleFields::Stage);
+        if (target) {
+          if (auto stage = target->optionalCast<CoilHeatingElectricMultiStageStageData>()) {
+            result.push_back(stage.get());
+          }
+        }
+      }
+      return result;
+    }
+
+    void CoilHeatingElectricMultiStage_Impl::addStage(CoilHeatingElectricMultiStageStageData& stage) {
+      auto group = getObject<ModelObject>().pushExtensibleGroup().cast<WorkspaceExtensibleGroup>();
+      group.setPointer(OS_Coil_Heating_Electric_MultiStageExtensibleFields::Stage, stage.handle());
+    }
+
+    boost::optional<HVACComponent> CoilHeatingElectricMultiStage_Impl::containingHVACComponent() const {
+      // AirLoopHVACUnitarySystem
+      std::vector<AirLoopHVACUnitarySystem> airLoopHVACUnitarySystems = this->model().getConcreteModelObjects<AirLoopHVACUnitarySystem>();
+
+      for (const auto& airLoopHVACUnitarySystem : airLoopHVACUnitarySystems) {
+        if (boost::optional<HVACComponent> heatingCoil = airLoopHVACUnitarySystem.heatingCoil()) {
+          if (heatingCoil->handle() == this->handle()) {
+            return airLoopHVACUnitarySystem;
+          }
+        }
+        if (boost::optional<HVACComponent> suppHeatingCoil = airLoopHVACUnitarySystem.supplementalHeatingCoil()) {
+          if (suppHeatingCoil->handle() == this->handle()) {
+            return airLoopHVACUnitarySystem;
+          }
+        }
+      }
+
+      // AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed
+      std::vector<AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed> airLoopHVACUnitaryHeatPumpAirToAirMultiSpeeds =
+        this->model().getConcreteModelObjects<AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed>();
+
+      for (const auto& airLoopHVACUnitaryHeatPumpAirToAirMultiSpeed : airLoopHVACUnitaryHeatPumpAirToAirMultiSpeeds) {
+        if (boost::optional<HVACComponent> heatingCoil = airLoopHVACUnitaryHeatPumpAirToAirMultiSpeed.heatingCoil()) {
+          if (heatingCoil->handle() == this->handle()) {
+            return airLoopHVACUnitaryHeatPumpAirToAirMultiSpeed;
+          }
+        }
+      }
+
+      return boost::none;
+    }
+
+    bool CoilHeatingElectricMultiStage_Impl::addToNode(Node& node) {
+      return false;
+    }
```

Comment:
Seems like we disallow putting the coil directly on an AirLoopHVAC branch (unlike E+)

Reply:
Following `CoilCoolingDXMultiSpeed` as a guide...

Reply:
Coil:Cooling:DX:MultiSpeed especially says so

```
Currently, this cooling coil can only be referenced by a [AirLoopHVAC:UnitarySystem](https://bigladdersoftware.com/epx/docs/22-1/input-output-reference/group-unitary-equipment.html#airloophvacunitarysystem) or [AirLoopHVAC:UnitaryHeatPump:AirToAir:Multispeed](https://bigladdersoftware.com/epx/docs/22-1/input-output-reference/group-unitary-equipment.html#airloophvacunitaryheatpumpairtoairmultispeed)
```

Reply:
I tried with an example file, it didn't work... seems like IDD /+ I/O freeze at out of line compared to the E+ code